Test/Tests/RealtimeClientPresenceTests.swift

```diff
@@ -756,86 +756,106 @@ class RealtimeClientPresenceTests: XCTestCase {
         let test = Test()
         let options = try AblyTests.commonAppSetup(for: test)
         let channelName = test.uniqueChannelName()
 
-        var clientMembers: ARTRealtime?
-        clientMembers = AblyTests.addMembersSequentiallyToChannel(channelName, members: 3, options: options)
-        defer { clientMembers?.dispose(); clientMembers?.close() }
-
-        options.clientId = "tester"
+        let clientMembers = AblyTests.addMembersSequentiallyToChannel(channelName, members: 2, options: options)
+        defer { clientMembers.dispose(); clientMembers.close() }
+        
+        options.clientId = "leaves"
+        let leavesClient = AblyTests.newRealtime(options).client
+        defer { leavesClient.dispose(); leavesClient.close() }
+        
+        options.clientId = "main"
         options.tokenDetails = try getTestTokenDetails(for: test, key: options.key!, clientId: options.clientId, ttl: 5.0)
-        let client = AblyTests.newRealtime(options).client
-        defer { client.dispose(); client.close() }
-        let channel = client.channels.get(channelName)
+        let mainClient = AblyTests.newRealtime(options).client
+        defer { mainClient.dispose(); mainClient.close() }
+        
+        let leavesChannel = leavesClient.channels.get(channelName)
+        let mainChannel = mainClient.channels.get(channelName)
+        
         waitUntil(timeout: testTimeout) { done in
-            let partialDone = AblyTests.splitDone(3, done: done)
-            channel.presence.enter(nil) { error in
+            let partialDone = AblyTests.splitDone(5, done: done)
+            mainChannel.presence.enter(nil) { error in
                 XCTAssertNil(error)
                 partialDone()
             }
-            channel.presence.get { members, error in
+            leavesChannel.presence.enter(nil) { error in
+                XCTAssertNil(error)
+                partialDone()
+            }
+            mainChannel.presence.get { members, error in
                 XCTAssertNil(error)
                 if let members {
-                    XCTAssertEqual(members.count, 3)
+                    XCTAssertEqual(members.count, 3) // "user1", "user2", "leaves"
                 } else {
                     XCTFail("Expected members to be non-nil")
                 }
                 partialDone()
             }
-            channel.presence.subscribe(.enter) { message in
-                if message.clientId == "tester" {
-                    channel.presence.unsubscribe()
+            mainChannel.presence.subscribe { message in
+                if message.clientId == "main" {
+                    XCTAssertEqual(message.action, ARTPresenceAction.enter)
+                    mainChannel.presence.unsubscribe()
+                    partialDone()
+                }
+                else if message.clientId == "leaves" {
+                    XCTAssertEqual(message.action, ARTPresenceAction.present) // .enter was replaced by .present (RTP2d)
```

Comment:
What do you mean ".enter was replaced by .present (RTP2d)"? RTP2d says "with an event name set to its original action" — there's nothing there to suggest that anything should be "replaced".

Reply:
Sorry, yes, I meant RTP2g.

What I'm saying is that as far as I can see there is nothing in the spec which says that `RealtimePresence` should "replace" the action before emitting it. RTP2d talks about what should be stored in the `PresenceMap`, not what should be emitted. And then RTP2g explicitly says that the received presence message should be emitted.

I think that what's probably happening here is timing-dependent, and that in the situations where you ran it, the `mainChannel.presence.enter()`’s side effects completed after the initial channel `SYNC`, and those of `leavesChannel.presence.enter()` after the initial `SYNC`, but I don't think this is guaranteed. I imagine that if you run this test multiple times, or if you put a delay between adding the presence subscriber and entering the channels, then you'll see that these events could be `PRESENT` or `ENTER`.

Whilst there are probably things you can do to get consistent behaviour here, I'd argue that it doesn't matter; you really just care that both members are present, so I'd suggest just checking (for both client IDs) that the action is `PRESENT` or `ENTER`.

Reply:
https://github.com/ably/ably-cocoa/pull/1793/commits/e20310d2c94bd8f870460e05973ac06dc6a44e15